processor/src/main/java/com/linecorp/decaton/processor/runtime/ProcessorSubscription.java

```diff
@@ -103,15 +113,63 @@ private void updateState(SubscriptionStateListener.State newState) {
                      .collect(Collectors.toSet());
     }
 
-    private void commitCompletedOffsets(Consumer<?, ?> consumer) {
+    // visible for testing
+    void commitCompletedOffsets(Consumer<?, ?> consumer, boolean sync) {
         Map<TopicPartition, OffsetAndMetadata> commitOffsets = contexts.commitOffsets();
         if (commitOffsets.isEmpty()) {
-            logger.debug("No new offsets to commit, skipping commit");
+            log.debug("No new offsets to commit, skipping commit");
             return;
         }
-        logger.debug("Committing offsets: {}", commitOffsets);
-        consumer.commitSync(commitOffsets);
-        contexts.updateCommittedOffsets(commitOffsets);
+        log.debug("Committing offsets {}: {}", sync ? "SYNC" : "ASYNC", commitOffsets);
+        if (sync) {
+            consumer.commitSync(commitOffsets);
+            contexts.updateCommittedOffsets(commitOffsets);
+        } else {
+            if (asyncCommitInFlight) {
+                // We would end up making multiple OffsetCommit requests containing same set of offsets if we proceed
+                // another async commit without waiting the first one to complete.
+                // This would be harmless if `decaton.commit.interval.ms` is set to sufficiently large, but otherwise
+                // we would make abusive offset commits despite there's no progress in processing.
+                log.debug("Skipping commit due to another async commit is currently in-flight");
+                return;
+            }
+            consumer.commitAsync(commitOffsets, (offsets, exception) -> {
+                asyncCommitInFlight = false;
+                if (exception != null) {
+                    log.warn("Offset commit failed asynchronously", exception);
+                    return;
+                }
+                // Exception raised from the commitAsync callback bubbles up to Consumer.poll() so it kills this
+                // subscription thread.
+                // Basically the exception thrown from here is considered a bug, but failing to commit offset itself
+                // isn't usually critical for consumer's continuation availability so we grab it here widely.
+                try {
```

Comment:
Hm, this try-catch block changes error handling semantics a bit from current impl.

current: `commitOffsetsIfNecessary` catches only `CommitFailedException | TimeoutException` so unexpected runtime exception from `contexts .updateCommittedOffsets` would kill subscription thread

And I'm not sure if it's good to catch widely here, since exception from updateCommittedOffset is completely unexpected (likely bug as you mentioned).

`updateCommittedOffset` just simply iterating over passed map and update each context's state, so exception from this method would be a sign of contexts state corruption which might be more critical?

Reply:
Good point. The only reason that I did this is because I thought likelihood of hitting bug is very different between sync (kind of procedural) and async case.
As we know the whole implementation of `updateCommittedOffsets` in case of sync we would almost never expect an exception to be raised, so that to me, it's kind of "really unexpected and have no idea what's going on" (so we should stop entire execution immediately).

> And I'm not sure if it's good to catch widely here, since exception from updateCommittedOffset is completely unexpected (likely bug as you mentioned).

The reason I'm trying to preserve execution even when we meet an unexpected exception here is explained in the above comment.

> updateCommittedOffset just simply iterating over passed map and update each context's state, so exception from this method would be a sign of contexts state corruption which might be more critical?

Exactly. However I think it still has some possibility to recover at next or few times later attempt because an exception possibly we see here is by accessing some missing element in contexts map that might happen possibly (by a bug) by timing issue rather than a permanent one. (e.g, missing context implies on next commit attempt the partition's offset won't be the subject of commit)